sql/alps_submitter.go

```diff
@@ -0,0 +1,167 @@
+package sql
+
+import (
+	"fmt"
+	"io"
+	"strconv"
+)
+
+type GraphConfig struct {
+	TableName string
+	Fields    []string
+	X         []FeatureSpec
+	Y         FeatureSpec
+}
+
+type FeatureSpec struct {
+	FeatureName string
+	IsSparse    bool
+	Shape       []int
+	DType       string
+}
+
+type NumericColumn struct {
+	Key   string
+	Shape int
+}
+
+type BucketColumn struct {
+	SourceColumn NumericColumn
+	Boundaries   []int
+}
+
+type CrossColumn struct {
+	Keys           []interface{}
+	HashBucketSize int
+}
+
+func resolveFeatureSpec(el *exprlist) (*FeatureSpec, error) {
+	headExpr := (*el)[0].val
+	if headExpr == "DENSE" {
+		shape, err := strconv.Atoi((*el)[2].val)
+		if err != nil {
+			return nil, err
+		}
+		return &FeatureSpec{
+			FeatureName: (*el)[1].val,
+			IsSparse:    false,
+			Shape:       []int{shape},
+			DType:       "double"}, nil
+	} else if headExpr == "SPARSE" {
+		shape, err := strconv.Atoi((*el)[2].val)
+		if err != nil {
+			return nil, err
+		}
+		return &FeatureSpec{
+			FeatureName: (*el)[1].val,
+			IsSparse:    false,
+			Shape:       []int{shape},
+			DType:       "double"}, nil
+	}
+	return nil, fmt.Errorf("not supported encoding type: %s", headExpr)
+}
+
+func resolveExpr(e *expr) (string, *exprlist) {
+	if e.val != "" {
+		return e.val, nil
+	} else {
+		return "", &e.sexp
+	}
+}
+
+func resolveExprList(el *exprlist) (interface{}, error) {
+	headName, headExprList := resolveExpr((*el)[0])
+	if headName == "" {
+		return resolveExprList(headExprList)
+	} else {
+		switch headName {
+		case "DENSE":
+			return resolveFeatureSpec(el)
+		case "SPARSE":
```

Comment:
"SPARSE" appears more than once. Perhaps we should define a const. (Same to "DENSE")

Reply:
Fixed